### PR TITLE
Create dockerfile to build image

### PR DESCRIPTION
**Pull request by trungdlp-wolffun**
Issue: https://github.com/hashicorp-forge/hermes/issues/8

### CODE REVIEW

File: .dockerignore [hunk 1]
```diff
@@ -0,0 +1 @@
+**/node_modules
```
Comment (inductor) [inline .dockerignore:1]: Add EOF in the end

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,37 @@
+# BUILD FRONTEND STAGE
```
Comment (inductor) [inline Dockerfile:1]:
```suggestion
# syntax = docker/dockerfile:1.4

# BUILD FRONTEND STAGE
```

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+FROM node:18-alpine as frontend-builder
```
Comment (inductor) [inline Dockerfile:2]:
```suggestion
FROM node:16-alpine as frontend-builder
```

File: Dockerfile [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+# Copy execute file
+WORKDIR /app
+COPY --from=backend-builder /app/hermes hermes
+
+# Set execuateable
+RUN chmod +x hermes
+
+# Execute app
+ENTRYPOINT ["./hermes"]
```
Comment (inductor) [inline Dockerfile:37]: Add EOF in the end

File: Dockerfile [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+# Install Make
+RUN apk add --no-cache make
+
+# Install Yarn v3
+RUN corepack enable
+RUN corepack prepare yarn@stable --activate
+RUN yarn set version stable
+
+# Build frontend
+RUN make web/build
```
Comment (inductor) [inline Dockerfile:17]:
```suggestion
WORKDIR /app

# Install Make
RUN apk add --no-cache make

# Install Yarn v3
RUN corepack enable
RUN corepack prepare yarn@stable --activate
RUN yarn set version stable

# Copy and install deps
COPY --link Makefile .
COPY --link web web
RUN make web/install-deps

# Copy all file
COPY --link . .

# Build frontend
RUN make web/build
```

File: Dockerfile [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+
+# BUILD BACKEND STAGE
+FROM golang:1.18 as backend-builder
+WORKDIR /app
+COPY --from=frontend-builder /app/web/dist /app/web/dist
+COPY . .
```
Comment (inductor) [inline Dockerfile:23]:
```suggestion
COPY --link go.mod go.sum ./
RUN go mod download
COPY --link . .
COPY --link --from=frontend-builder /app/web/dist /app/web/dist
```

**Comment by inductor** [comment]
@trungdlp-wolffun added some Dockerfile improvements including latest specification in https://docs.docker.com/engine/reference/builder/#copy---link

File: Dockerfile (hunk 3 above)
Comment (trungdlp-wolffun) [reply Dockerfile:2]: Why we should use Node.js 16?

**Comment by trungdlp-wolffun** [comment]
@inductor I'm updated your suggestion. Can you help me understand `COPY --link` meaning ? (I already read Docker document)

**Comment by inductor** [comment]
@trungdlp-wolffun https://github.com/moby/buildkit/blob/master/frontend/dockerfile/docs/reference.md#copy---link would probably have a better explanation.

It has better caching than without it

File: Dockerfile [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+FROM golang:1.18 as backend-builder
+WORKDIR /app
+COPY --link go.mod go.sum ./
+RUN go mod download
+COPY --link . .
+COPY --link --from=frontend-builder /app/web/dist /app/web/dist
+RUN make bin
+
+# RUN STAGE
+FROM alpine:latest
```
Comment (inductor) [inline Dockerfile:36]: I would avoid using latest tag and might adopt something like distroless base image. But I would like project maintainers to make this decision

Comment (trungdlp-wolffun) [reply Dockerfile:36]: @inductor I think distroless better than alpine, smaller and only necessary dependency.

Comment (trungdlp-wolffun) [reply Dockerfile]: I updated alpine to ver 3.17

Comment (trungdlp-wolffun) [reply Dockerfile]: @inductor Do you know https://www.chainguard.dev/chainguard-images?

**Comment by sjbarag** [comment]
Any updates here? Having a Dockerfile upstream would make Hermes much easier to deploy!

**Comment by jeffdaley** [comment]
Hey all, thanks for the contributions.

Unfortunately we're not accepting Dockerfile PRs right now, but this is [planned work](https://github.com/hashicorp-forge/hermes/issues?q=is%3Aissue+is%3Aopen+label%3Aplanned-work). Basically, we have internal, build-related things to configure first. 

We appreciate your continued patience on this.

**Pull request closed without merging**